lardata/ArtDataHelper/Dumpers/DumpOpFlashes_module.cc

```diff
@@ -0,0 +1,253 @@
+    using Comment = fhicl::Comment;
+
+    fhicl::Atom<art::InputTag> OpFlashModuleLabel{
+      Name{"OpFlashModuleLabel"},
+      Comment{"tag of the recob::OpFlash collection to be dumped"}};
+
+    fhicl::Atom<bool> PrintOpHitAssociations{
+      Name("PrintOpHitAssociations"),
+      Comment("also prints a list of optical hits associated to the flash"),
+      true};
+
+    fhicl::Atom<std::string> OutputCategory{
+      Name("OutputCategory"),
+      Comment("the messagefacility category used for the output"),
+      "DumpOpFlashes"};
+
+  }; // Config
+
+  using Parameters = art::EDAnalyzer::Table<Config>;
+
+  /// Default constructor
+  explicit DumpOpFlashes(Parameters const& config);
+
+  /// Does the printing
+  void analyze(const art::Event& evt) override;
+
+private:
+  art::InputTag const fOpFlashModuleTag; ///< Optical hit data product tag.
+  bool const fPrintOpHitAssociations;    ///< Whether to print optical hits.
+  std::string const fOutputCategory;     ///< Category for `mf::LogInfo` output.
+
+}; // class ophit::DumpOpFlashes
+
+// -----------------------------------------------------------------------------
+// ---  implementation
+// -----------------------------------------------------------------------------
+namespace {
+
+  struct ProductSourceEntry_t {
+    std::size_t index = 0;                        ///< An incremental index.
+    art::BranchDescription const* info = nullptr; ///< Pointer to the product information.
+  };
+  using ProductSourceDirectory_t = std::unordered_map<art::ProductID, ProductSourceEntry_t>;
+
+  /**
+   * @brief Returns a map from product ID to product information for all
+   *        pointers on one side of an association.
+   * @tparam Side which side of the association (either `0` or `1`)
+   * @tparam Assns the type of association
+   * @tparam Event the type of event (makes this code gallery-compatible)
+   * @param assns the association to be parsed
+   * @param event the event holding the information
+   * @return a map from product ID to product information
+   *
+   * The product information is a pair:
+   *  * `first`: an incremental index
+   *  * `second`: a pointer to the product information (_art_ branch description)
+   */
+  template <std::size_t Side = 1U, typename Assns, typename Event>
+  ProductSourceDirectory_t buildAssnsSourceDirectory(Assns const& assns, Event const& event)
+  {
+
+    ProductSourceDirectory_t dir;
+
+    // extract all the unique IDs
+    for (auto const& ptrs : assns)
+      dir.emplace(std::get<Side>(ptrs).id(), ProductSourceEntry_t{});
+
+    // assign the index and pointer to each entry (order is not defined)
+    std::size_t index{0};
+    for (auto& [id, info] : dir)
+      info = {index++, event.getProductDescription(id).get()};
+
+    return dir;
+  } // buildAssnsSourceDirectory()
+
+  //----------------------------------------------------------------------------
+  template <typename Stream>
+  void dumpAssociatedOpHits(Stream& out,
```

Comment:
Suggest not using a template as only one instantiation is required—I'd replace `Stream` with `mf::LogVerbatim`.  I'd also remove all default arguments as the function is called at only one call site.

Reply:
I do not like to hard-code a "specific logger class" in the interface. If there were a base class to pass I would use that.
I can code it for `std::ostream` instead; but I don't know a simple way to do that, so I have used a wrapper.

Reply:
> I do not like to hard-code a "specific logger class" in the interface. If there were a base class to pass I would use that.

Why?  We hard-code specific types all of the time (e.g. `std::string`).  It only makes sense to make a template when more than one type needs to serve as a template argument.  But only one type is used in this translation unit...and other translation units do not have access to this function template.  Making this a template is thus overkill and premature generalization.